### PR TITLE
Functionality to insert aggregated data into graphite.

### PR DESCRIPTION
**Pull request by rishubhjain**
tendrl-bug-id: Tendrl/monitoring-integration#32

### CODE REVIEW

File: etc/tendrl/monitoring-integration/monitoring-integration.conf.yaml.sample [hunk 1]
```diff
@@ -29,6 +29,10 @@ datasource_name: 'Graphite'
 
 datasource_host: 127.0.0.1
 
-access: direct
+datasource_port: 10080
+
+carbon_port: 2003
+
```
Comment (r0h4n) [inline etc/tendrl/monitoring-integration/monitoring-integration.conf.yaml.sample]: Add help comments for each item, example,  access: proxy (I am not able to understand the purpose of flag without help comment)

File: tendrl/monitoring_integration/sync/__init__.py [hunk 2]
```diff
@@ -11,6 +16,28 @@ def __init__(self):
...
+        except ValueError as ex:
+            logger.log("error", NS.get("publisher_id", None),
+                       {'message': str(ex)})
+            raise ex
+
+        while not self._complete.is_set():
+            try:
+                gevent.sleep(interval)
+                plugin_obj = GraphitePlugin()
+                objects  = NS.monitoring.definitions.get_parsed_defs()["namespace.monitoring"]["graphite_data"]
```
Comment (r0h4n) [inline tendrl/monitoring_integration/sync/__init__.py]: These "objects" arent going to change since we dont yet allow updates to definitions of a live Tendrl cluster, so please move "objects = .."  outside the while loop

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
+import os
+import copy
+
+
+import etcd
+import time
+import urllib3
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: urllib3 rpm available for epel and rhel  right?

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    def __init__(self):
+
+        self.host = NS.config.data["datasource_host"]
+        self.port = NS.config.data["datasource_port"]
+        self.carbon_port = NS.config.data["carbon_port"]
+
+        self.graphite_sock = socket.socket()
+        self.graphite_sock.connect((self.host, int(self.carbon_port)))
+        self.http = urllib3.PoolManager()
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: self.http is not used anywhere, why is it required? Is urllib even required?

File: tendrl/monitoring_integration/sync/__init__.py [hunk 5]
```diff
@@ -11,6 +16,28 @@ def __init__(self):
...
+
+        except ValueError as ex:
+            logger.log("error", NS.get("publisher_id", None),
+                       {'message': str(ex)})
+            raise ex
+
+        while not self._complete.is_set():
+            try:
+                gevent.sleep(interval)
+                plugin_obj = GraphitePlugin()
```
Comment (r0h4n) [inline tendrl/monitoring_integration/sync/__init__.py]: Do you really need to create instance of GraphitePlugin every iteration? Move it outside the while loop, and in the GraphitePlugin class handle any socket read/write errors

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    def push_metrics(self, metric_name, metric_value):
+        message = '%s%s%s %s %d\n' % (
+            self.prefix,
+            str("."),
+            metric_name,
+            str(metric_value),
+            int(time.time())
+        )
+
+        response = self.graphite_sock.sendall(message)
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: catch socket errors and refresh socket connection if required

File: tendrl/monitoring_integration/grafana/utils.py [hunk 7]
```diff
@@ -16,9 +16,8 @@ def get_conf():
     try:
 
         # Graphite and Grafana will be running on localhost
-        NS.config.data["grafana_host"] = "localhost"
+        NS.config.data["grafana_host"] = "127.0.0.1"
```
Comment (r0h4n) [inline tendrl/monitoring_integration/grafana/utils.py]: why is this change required?

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
+import os
+import copy
+
+
+import etcd
+import time
+import socket
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: Use gevent.socket instead of just socket

http://www.gevent.org/gevent.socket.html

It is a drop in replacement of socket. Same functions to send and receive

File: tendrl/monitoring_integration/sync/__init__.py [hunk 9]
```diff
@@ -11,6 +16,28 @@ def __init__(self):
...
+        try:
+            interval = float(interval.value)
+        except ValueError as ex:
+            logger.log("error", NS.get("publisher_id", None),
+                       {'message': str(ex)})
+            raise ex
 
+        
+        plugin_obj = GraphitePlugin()
+        objects  = NS.monitoring.definitions.get_parsed_defs()["namespace.monitoring"]["graphite_data"]
```
Comment (r0h4n) [inline tendrl/monitoring_integration/sync/__init__.py]: Instead of objects. Call it aggregate_gluster_objects

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        try:
+            if obj_attr["count"]:
+                self.get_count(resource_details, obj_attr)
+        except KeyError:
+            pass
+        return resource_details
+                 
+
+    def get_data(self, objects):
+        try:
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: Call this get_central_store_data

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+                           up = up + 1
+                if key == "down":
+                   for attr_key, attr_values in obj_attr["count"]["down"].items():
+                       if resource_detail[attr_key] in attr_values:
+                           down = down + 1
+        resource_details["total"] = total
+        resource_details["up"] = up
+        resource_details["down"] = down
+
+    def read_data(self, resource_key, obj_attr):
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: Call this get_object_from_central_store

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+            metric_name,
+            str(metric_value),
+            int(time.time())
+        )
+        
+        response = self.graphite_sock.sendall(message)
+        return response
+            
+
+    def get_count(self,resource_details, obj_attr):
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: Call this get_resource_count

File: tendrl/monitoring_integration/graphite/__init__.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+        message = '%s%s%s %s %d\n' % (
+            self.prefix,
+            str("."),
+            metric_name,
+            str(metric_value),
+            int(time.time())
+        )
+        
+        response = self.graphite_sock.sendall(message)
```
Comment (r0h4n) [inline tendrl/monitoring_integration/graphite/__init__.py]: Catch errors in sendall and , if connection issue, retry sending message again just once

File: tendrl/monitoring_integration/grafana/utils.py (hunk 7 above)
Comment (rishubhjain) [inline tendrl/monitoring_integration/grafana/utils.py]: @r0h4n localhost can be mapped to some other ip, so to be specific I have mentioned 127.0.0.1 instead of localhost
If this is not needed then I would revert it back
